climakitae/explore/vulnerability.py

```diff
@@ -25,27 +40,72 @@ def metric_agg(da, metric, threshold):
     metric : str
         The metric type to calculate. Supported options are "min", "max",
         "mean", and "median".
-    threshold : float
-        The threshold value for comparison.
+    heat_idx_threshold : float
+        The heat index threshold for comparison.
 
     Returns:
     --------
     metric_data : xarray.DataArray
-        Resulting DataArray from the calculated metric and threshold applied.
+        Resulting DataArray from the calculated metric and heat index threshold applied.
     """
     metric_map = {
         "min": np.min,
         "max": np.max,
-        "mean": np.mean,
-        "median": np.median,
     }
 
-    metric_data = (
-        (da.resample(time="1D").reduce(metric_map[metric]) > threshold)
-        .resample(time="1Y")
-        .sum()
-    )
+    def apply_metric(da):
+        """Applies the hard-coded heat index threshold or percentile to the DataArray"""
+        if (
+            heat_idx_threshold or heat_idx_threshold == 0
+        ):  # If a threshold is passed and/or if the `heat_idx_threshold` passed in is 0, since that is a valid threshold.
+            calc_val = (
+                (da.resample(time="1D").reduce(metric_map[metric]) > heat_idx_threshold)
+                .resample(time="1Y")
+                .sum()
+                .mean(dim="time")
+            )  # This line counts the average number of high heat days per year over this time interval
+            calc_val.name = f"Days per year with {metric.capitalize()} Daily {da.name} above a heat index threshold of {heat_idx_threshold} {da.units}"
+            return calc_val
+
+        elif one_in_x:  # Calculate 1-in-X events
+
+            # Get 1-in-x event values
+            event_vals = []
+            for (
+                sim
+            ) in (
+                da.simulation.values
+            ):  # Can use xr.apply_ufunc, not just a for loop here
+                event_val = get_block_maxima(  # Victoria: is this function right and/or am I using it right? I think these counts are a bit low.
+                    da.sel(simulation=sim),
+                    duration=(24, "hour"),
+                    extremes_type=metric,
+                    block_size=one_in_x,
+                ).reduce(
+                    np.mean
+                )
+                event_vals.append(event_val)
```

Comment:
Okay, I think I have figured this out! 

```suggestion
        elif one_in_x:  # Calculate 1-in-X events

            # Get 1-in-x event values
            event_vals = []
            for (
                sim
            ) in (
                da.simulation.values
            ):  # Can use xr.apply_ufunc, not just a for loop here
                event_val = get_block_maxima( 
                    da.sel(simulation=sim),
                    extremes_type=metric,
                    groupby=(1, 'day'),
                ).reduce(
                    np.mean
                )
                event_vals.append(event_val)
```

This first step calculates the annual maximum series which goes into the `get_return_value` function (this was missing!) Something like:
```
return_values = get_return_value(DATAARRAY_FROM_ABOVE, return_period=1_in_x, multiple_points=False)
```

Now this didn't quite work for me in testing because of the following error:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [28], line 1
----> 1 test_1_in_x = get_return_value(ams1, return_period=10, multiple_points=False)
      2 test_1_in_x

File ~/src/climakitae/climakitae/explore/threshold_tools.py:737, in get_return_value(bms, return_period, distr, bootstrap_runs, conf_int_lower_bound, conf_int_upper_bound, multiple_points)
    703 def get_return_value(
    704     bms,
    705     return_period=10,
   (...)
    710     multiple_points=True,
    711 ):
    712     """Creates xarray Dataset with return values and confidence intervals from maximum series.
    713 
    714     Parameters
   (...)
    734         Dataset with return values and confidence intervals
    735     """
--> 737     return _get_return_variable(
    738         bms,
    739         "return_value",
    740         return_period,
    741         distr,
    742         bootstrap_runs,
    743         conf_int_lower_bound,
    744         conf_int_upper_bound,
    745         multiple_points,
    746     )

File ~/src/climakitae/climakitae/explore/threshold_tools.py:667, in _get_return_variable(bms, data_variable, arg_value, distr, bootstrap_runs, conf_int_lower_bound, conf_int_upper_bound, multiple_points)
    657     return return_variable, conf_int_lower_limit, conf_int_upper_limit
    659 return_variable, conf_int_lower_limit, conf_int_upper_limit = xr.apply_ufunc(
    660     _return_variable,
    661     bms,
   (...)
    664     output_core_dims=[[], [], []],
    665 )
--> 667 return_variable = return_variable.rename(data_variable)
    668 new_ds = return_variable.to_dataset()
    669 new_ds["conf_int_lower_limit"] = conf_int_lower_limit

File /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/dataset.py:3584, in Dataset.rename(self, name_dict, **names)
   3556 def rename(
   3557     self: T_Dataset,
   3558     name_dict: Mapping[Any, Hashable] | None = None,
   3559     **names: Hashable,
   3560 ) -> T_Dataset:
   3561     """Returns a new object with renamed variables, coordinates and dimensions.
   3562 
   3563     Parameters
   (...)
   3582     DataArray.rename
   3583     """
-> 3584     name_dict = either_dict_or_kwargs(name_dict, names, "rename")
   3585     for k in name_dict.keys():
   3586         if k not in self and k not in self.dims:

File /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/utils.py:284, in either_dict_or_kwargs(pos_kwargs, kw_kwargs, func_name)
    281     return cast(Mapping[Hashable, T], kw_kwargs)
    283 if not is_dict_like(pos_kwargs):
--> 284     raise ValueError(f"the first argument to .{func_name} must be a dictionary")
    285 if kw_kwargs:
    286     raise ValueError(
    287         f"cannot specify both keyword and positional arguments to .{func_name}"
    288     )

ValueError: the first argument to .rename must be a dictionary
```